LUCENE-9932: Performance improvement for BKD index building

neoremind opened the pull request:
https://issues.apache.org/jira/browse/LUCENE-9932

# Description

In BKD index building, the input bytes must be sorted before calling BKD writer related API. The sorting method leverages MSB Radix Sort algorithm, and the comparing method takes both the bytes itself and the DocId, but in real cases, DocIds are usually monotonically increasing. This could yield one possible performance enhancer. I found this enhancement when I dig into one performance issue in our system. Then I research on the possible solution.

# Solution

DocId is usually increased by one when building index in a thread-safe way, by assuming such condition, the comparing method can eliminate the unnecessary comparing input - DocId, only leave the bytes itself to compare. In order to do so, MSB radix sorting and its fallback sorting method must be **stable**, so that when elements are the same, the sorting method maintains its original order when added, which makes DocId still monotonically increasing. To make MSB Radix Sort stable, it needs a trivial update; to make fallback sort table, use merge sort instead of quick sort. Meanwhile, there should introduce a switch which is able to turn the stable option on or off.

# Tests

Add testcases.

Add performance testcase (maybe removed when merging).

Test environment:
MacBook Pro (Retina, 15-inch, Mid 2015), 2.2 GHz Intel Core i7, 16 GB 1600 MHz DDR3

Java version:
java version "1.8.0_161"
Java(TM) SE Runtime Environment (build 1.8.0_161-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.161-b12, mixed mode)

Test scenarios:
bytesPerDim = [1, 2, 3, 4, 8, 16, 32]
dim = 1
doc num = 2,000,000
warm up 5 time, run 10 times to calculate average time used.

bytesPerDim\scenario | disable sort doc id (PR branch) | enable sort doc id (master branch)
-- | -- | --
1 | 30989.594 us | 1151149.9 us
2 | 313469.47 us | 1115595.1 us
3 | 844617.8 us | 1465465.1 us
4 | 1350946.8 us | 1465465.1 us
8 | 1344814.6 us | 1458115.5 us
16 | 1344516.6 us | 1459849.6 us
32 | 1386847.8 us | 1583097.5 us

Result shows that, by disabling sort DocId, sorting runs 1.73x to 37x faster when there are many duplicate bytes (bytesPerDim = 1 or 2 or 3). When data cardinality is high (bytesPerDim >= 4, test cases will generate random bytes which are more scatter, not likely to be duplicate), the performance does not go backward, still a little better.

In conclusion, in the end to end process for building BKD index, which relies on BKDWriter for some data types, performance could be better by ignoring DocId if they are already monotonically increasing.

neoremind committed:
use stable sort in msb radix sort

jpountz commented:
Wow, this is a neat idea! I wonder if we could find a less intrusive way to apply this change, e.g. could we avoid adding info about whether to sort by doc ID to BKDConfig? For instance could MSBRadixSorter detect whether data is sorted by itself instead of having to propagate this information in so many places?

neoremind commented:
@jpountz Good advice! Before that I am still struggling where to propagate this config up to the index builder layer. 
I will give it a try, the first thing comes up my mind is to bring up a new `prepare` method, in which it will scan all docid from i to j to see if they are increasing. I will experiment on this, if the overhead is small enough, then it is worthwhile to sort without docid.
One more question, are there any places where doc id is not added increasingly? I mean the source code, not test cases.

jpountz commented:
> I will give it a try, the first thing comes up my mind is to bring up a new prepare method,

One idea I had in mind was to create a new class, something like `StableMSBRadixSorter` that would extend `MSBRadixSorter` to:
 - add the two `assign` and `finalizeAssign` methods that you currently added to `Sorter`,
 - override the way data gets rearranged to guarantee stability,
 - change the fallback sorter,
 - modify `radixSort(int,int,int,int)` to check whether data is already sorted before computing the common prefix length and the histogram of the leading bytes.

> One more question, are there any places where doc id is not added increasingly?

I don't remember how we deal with it, but we should check how this optimization plays with index sorting, since we would renumber doc IDs at flush time.

jpountz commented on lucene/core/src/java/org/apache/lucene/util/bkd/MutablePointsReaderUtils.java:
```diff
@@ -39,13 +41,23 @@
   public static void sort(
       BKDConfig config, int maxDoc, MutablePointValues reader, int from, int to) {
     final int bitsPerDocId = PackedInts.bitsRequired(maxDoc - 1);
```
I wonder if we could check whether doc IDs are sorted right above this line, and then adapt the number of compared bytes? Then we should be able to get rid of `isEnableStableSort` and `doClosureIfStableSortEnabled`?

```suggestion
    // In the common case that index sorting is not enabled, doc IDs come in order, which enables us to check fewer bytes for sorting
    boolean sortedDocIDs = true;
    for (int i = from + 1; i < to; i++) {
      if (reader.getDocID(i - 1) > reader.getDocID(i)) {
        sortedDocIDs = false;
        break;
      }
    }
    final int bitsPerDocId = sorted ? 0 : PackedInts.bitsRequired(maxDoc - 1);
```

neoremind replied:
Is this a little bit better?
```
bool canUseStableSort = check....
if (canUseStableSort) {
  StableMsbRadixSort.sort(..)
} else {
  MsbRadixSort.sort(..)
}
```

jpountz replied:
Good question, my gut feeling is that the benefits of switching between the stable and the unstable radix sort are not worth the code complexity. Let's always use the stable sort?

neoremind replied:
Will refine after we decide whether to always use stable sort.

neoremind commented:
@jpountz Per your advice, I have updated the code.

In terms of performance, I refined `TestBKDDisableSortDocId`, to make it re-runnable as a benchmark. I have made the following benchmark.

doc num = 2,000,000, dim number = 1, bytePerDim = [1,2,3,4,8,16,32], run times = 10 (warm up 5 times)

Result is shown as below.

```
  -------------------------------------------------
| bytesPerDim | isDocIdIncremental | avg time(us) |
 -------------------------------------------------
|      1      |         N          |   1127688.5  |
|      1      |         Y          |     56464.7  |
|      2      |         N          |   1124137.8  |
|      2      |         Y          |    339150.2  |
|      3      |         N          |   1485020.4  |
|      3      |         Y          |    878251.1  |
|      4      |         N          |   1436003.9  |
|      4      |         Y          |   1376974.1  |
|      8      |         N          |   1444971.3  |
|      8      |         Y          |   1365877.5  |
|     16      |         N          |   1500235.5  |
|     16      |         Y          |   1385235.8  |
|     32      |         N          |   1516514.0  |
|     32      |         Y          |   1415364.9  |
 -------------------------------------------------
```

Meanwhile, I also reset to main branch to run the same test case. Result is shown as below.

```
  -------------------------------------------------
| bytesPerDim | isDocIdIncremental | avg time(us) |
 -------------------------------------------------
|      1      |         N          |   1144138.9  |
|      1      |         Y          |    390398.8  |
|      2      |         N          |   1121301.9  |
|      2      |         Y          |   1124444.9  |
|      3      |         N          |   1482451.3  |
|      3      |         Y          |   1471338.1  |
|      4      |         N          |   1424961.6  |
|      4      |         Y          |   1423907.1  |
|      8      |         N          |   1437768.5  |
|      8      |         Y          |   1474001.1  |
|     16      |         N          |   1464370.4  |
|     16      |         Y          |   1478291.3  |
|     32      |         N          |   1500323.9  |
|     32      |         Y          |   1508646.1  |
 -------------------------------------------------
```

I made a graph so that we can see more clearly.

![](https://issues.apache.org/jira/secure/attachment/13024383/refined-code-benchmark.png)

If DocIds are increasing, PR branch out-performs in all scenarios. 
If DocIds are not in order, we expect the performance to be the same with main branch. It does work almost the same, but here we introduce a small overhead to scan data beforehand, checking whether data is in order, so PR branch is a little bit (like 1% percent) backward. 

I made a flame-graph, the right-most column is where checking order consumes, very small, like below 3% of total CPU consumption.
![](https://issues.apache.org/jira/secure/attachment/13024385/flame-graph.png)

neoremind commented on lucene/core/src/java/org/apache/lucene/util/StableMSBRadixSorter.java:
```diff
@@ -0,0 +1,91 @@
+      stableReorder(from, to, startOffsets, endOffsets, k);
+    } else {
+      super.reorder(from, to, startOffsets, endOffsets, k);
+    }
```
Good catch. If I override the method, then the parent class will use stable reorder. Interestingly, I found non-stable sorting runs faster accidentally (I debug the whole day :joy:). So to make the behavior work as expected, I update the code in such way.

neoremind replied:
There is no silver bullet, the stable reorder runs faster, I think the reason is that it makes less swapping, but using extra space to store docid.

neoremind commented on lucene/core/src/java/org/apache/lucene/util/StableMSBRadixSorter.java:
```diff
@@ -0,0 +1,91 @@
+  /** Finalize assign operation, to switch array. */
+  protected void finalizeAssign(int from, int to) {
+    throw new UnsupportedOperationException("not implement");
+  }
```
ok, I will.

neoremind replied:
done.

neoremind commented on lucene/core/src/java/org/apache/lucene/codecs/MutablePointValues.java:
```diff
@@ -41,4 +41,14 @@ protected MutablePointValues() {}
+  /** Finalize assign operation, to switch array. */
+  public void finalizeAssign(int from, int to) {
+    throw new UnsupportedOperationException();
+  }
```
will do.

jpountz commented:
I left a few comments before your previous comment, I left some suggestions to make the code simpler but I think that it's getting close.

Thanks for the great performance analysis! Given that you are benchmarking with fully random bytes, I'm actually expecting the speedup to be greater with real data where the leading bytes are often zeroes. For instance I'd expect users who index integers (4 bytes) between 0 and 2^24 to notice speedups that are closer to the one that you computed for bytesPerDim=3 than for bytesPerDim=4.

iverase commented:
What I read for this results is that the stable sorting is faster than the non-stable. I think it makes sense because re-ordering using a helper array is probably faster that in-place re-ordering. 

If we are ok to take the memory penalty of using this extra array, then I think it make sense to always use the stable version of the algorithm?

neoremind commented:
> For instance I'd expect users who index integers (4 bytes) between 0 and 2^24 to notice speedups that are closer to the one that you computed for bytesPerDim=3 than for bytesPerDim=4.
I can also construct a real case to verify :smile:

jpountz commented:
> If we are ok to take the memory penalty of using this extra array, then I think it make sense to always use the stable version of the algorithm?

+1 to always use the stable version of the algorithm. This would only use transient memory and in reasonable amounts, so I'm not concerned with the memory usage.

neoremind committed:
make some refinement per comment, make stable order as default strategy.

neoremind commented:
> +1 to always use the stable version of the algorithm. This would only use transient memory and in reasonable amounts, so I'm not concerned with the memory usage.

Per comment, I make stable reorder in MSB Radix Sort as default. Things become a little tricky.

The benchmark of this PR branch goes like below. Starting from bytePerDim=4, non-stable MSB Radix Sort runs faster than stable version.
```
  -------------------------------------------------
| bytesPerDim | isDocIdIncremental | avg time(us) |
 -------------------------------------------------
|      1      |         N          |    981732.7  |
|      1      |         Y          |     56766.1  |
|      2      |         N          |    941666.1  |
|      2      |         Y          |    304395.9  |
|      3      |         N          |   1306374.9  |
|      3      |         Y          |    853744.8  |
|      4      |         N          |   1262810.3  |
|      4      |         Y          |   1345605.8  |
|      8      |         N          |   1282193.9  |
|      8      |         Y          |   1364078.3  |
|     16      |         N          |   1269967.4  |
|     16      |         Y          |   1433378.3  |
|     32      |         N          |   1293686.0  |
|     32      |         Y          |   1468472.1  |
 -------------------------------------------------
```

To make it more clear, I also add the new result into the graph I made above.

![](https://issues.apache.org/jira/secure/attachment/13024398/refined-code-benchmark2.png)

I try to analysis, according to the flame graph I generated (see above picture). Sorting includes the following stages.
- build histogram
- fallback intro sort (quick sort for non-stable, in place merge sort for stable)
- reorder

Let's illustrate the total time partitioned by the cost per stage. 

For main branch `MSBRadixSort`, it shows as below.
```
+--------------------------------------------------------------------------------------------------------+
|  build histogram |      fallback sort (quick sort)     |             non stable reorder                | 
+--------------------------------------------------------------------------------------------------------+   
```

For PR branch `StableMSBRadixSort`, it shows as below
```
+-------------------------------------------------------------------------------------------------+
|  build histogram |         fallback sort (in place merge sort)        |    stable reorder       | 
+-------------------------------------------------------------------------------------------------+ 
```

For PR branch `MSBRadixSort` making stable reorder as default, it shows as below. 
```
+----------------------------------------------------------------------------------+
|  build histogram |       fallback sort (quick sort)     |    stable reorder      | 
+----------------------------------------------------------------------------------+   
```

I think that stable reorder contributes most part of the speedup, but in place merge sort is slower than 3-pivot quick sort.

After all, regardless of the order of doc IDs, stable reorder is definitely a good choice. If data are spread evenly, not many duplicates, `StableMSBRadixSort` is a little slower (10%) than `MSBRadixSorter`. If not, `StableMSBRadixSort` will out-perform far better than `MSBRadixSorter`. This makes me reluctant on the way we should go, @jpountz could you give me some advice?

neoremind commented:
> For instance I'd expect users who index integers (4 bytes) between 0 and 2^24 to notice speedups that are closer to the one that you computed for bytesPerDim=3 than for bytesPerDim=4.

@jpountz I add a new benchmark test case `testBenchmarkWithLeadingZeroBytes` in `TestBKDDisableSortDocId`. Your assumption is correct.

Result is shown as below (do not make stable reorder as default).
```
bytesPerDim=4, leadingZeroByteNum=[1,2,3]
 -------------------------------------------------
| bytesPerDim | isDocIdIncremental | avg time(us) |
 -------------------------------------------------
|      4      |         N          |   1476874.0  |  <- leadingZeroByteNum = 1
|      4      |         Y          |   1363280.3  | <- leadingZeroByteNum = 1
|      4      |         N          |   1601885.9  | <- leadingZeroByteNum = 2
|      4      |         Y          |    930728.1  | <- <- leadingZeroByteNum = 2
|      4      |         N          |   1210586.8  | <- leadingZeroByteNum = 3
|      4      |         Y          |    367141.6  | <- <- leadingZeroByteNum = 3
 -------------------------------------------------
```

jpountz commented:
Very cool, thanks for testing what happens with small integers!

Do the numbers get better if you use `TimSorter` as a fallback sort? `InPlaceMergeSorter` is convenient because it only requires implementing a few methods, but the logic it uses to merge sorted arrays in-place is very slow, so I wonder if that's the cause of the slow down we're seeing.

neoremind commented:
I use `TimSort` instead of `InPlaceMergeSorter`, expect it to be faster, but it turns out to be slower. @jpountz would you check my latest commit to see if I implement Tim Sort correctly? 

Below is the latest benchmark of `MSBRadixSort` with stable reorder(isDocIdIncremental = N) and `StableMSBRadixSort` (isDocIdIncremental = Y) 
```
 -------------------------------------------------
| bytesPerDim | isDocIdIncremental | avg time(us) |
 -------------------------------------------------
|      1      |         N          |    995541.5  |
|      1      |         Y          |     60399.2  |
|      2      |         N          |    951085.9  |
|      2      |         Y          |    322054.3  |
|      3      |         N          |   1333992.5  |
|      3      |         Y          |    756951.4  |
|      4      |         N          |   1340422.4  |
|      4      |         Y          |   1528955.5  |
|      8      |         N          |   1323878.8  |
|      8      |         Y          |   1494004.5  |
|     16      |         N          |   1305548.1  |
|     16      |         Y          |   1480329.4  |
|     32      |         N          |   1326447.5  |
|     32      |         Y          |   1589089.8  |
 -------------------------------------------------
```

jpountz commented:
Oh, this is disappointing, maybe the arrays are too small for TimSorter to actually perform better than InPlaceMergeSorter.

I'd be keen to proceed with the change that always performs a stable sort with InPlaceMergeSorter. Some cases do get slower but only by a few percents, and it's going to be unlikely noticed through the full indexing chain. On the other hand, some cases are getting several times faster, which I'm sure is going to be noticeable. We could still iterate later, but for now this sounds to me like a good performance-simplicity trade-off. What do you think?

neoremind committed:
make stable radix sorter as default sorter

neoremind commented:
I make `StableMSBRadixSorter` as default sorter and use `InPlaceMergeSorter` as fallback sorter. Please check the latest commit. (I did not squash commits, and save every commit as separate branch in my own forked repo for future reference).

Average sort time (unit in us) comparison.
(1) PR branch
(2) main branch
(3) PR branch, verify doc ID not in ascending order, original `MSBRadixSorter` with stable reorder.
```
 -------------------------------------------------------------------------------
| bytesPerDim | isDocIdIncremental |      (1)     |      (2)     |     (3)     |
 -------------------------------------------------------------------------------
|      1      |         N          |     31132.7  |   1144138.9  |    981732.7  |
|      1      |         Y          |     33049.8  |    390398.8  |              |
|      2      |         N          |    273975.1  |   1121301.9  |    941666.1  |
|      2      |         Y          |    281060.2  |   1124444.9  |              |
|      3      |         N          |    844291.7  |   1482451.3  |   1306374.9  |
|      3      |         Y          |    804839.8  |   1471338.1  |              |
|      4      |         N          |   1274670.8  |   1424961.6  |   1262810.3  |
|      4      |         Y          |   1289128.8  |   1423907.1  |              |
|      8      |         N          |   1357592.3  |   1437768.5  |   1282193.9  |
|      8      |         Y          |   1286732.0  |   1474001.1  |              |
|     16      |         N          |   1366177.6  |   1464370.4  |   1269967.4  |
|     16      |         Y          |   1353213.1  |   1478291.3  |              |
|     32      |         N          |   1403655.4  |   1500323.9  |   1293686.0  |
|     32      |         Y          |   1406872.4  |   1508646.1  |              |
 -------------------------------------------------------------------------------
```

In conclusion, PR branch runs faster than main branch regardless of bytePerDim and isDocIdIncremental, some cases 38x times faster. If we make original `MSBRadixSorter` with stable reorder, PR branch is little bit slower but acceptable. I think in real case, doc ID is usually in ascending order when index building, and that could be a noticeable performance improvement.

> We could still iterate later, but for now this sounds to me like a good performance-simplicity trade-off. What do you think?
I have no problem 😄

neoremind commented:
@jpountz are there any testcases suitable to verify the end to end performance improvement, like through IndexWriter? maybe I could give it try.

jpountz commented:
Maybe try running this benchmark: https://github.com/mikemccand/luceneutil/blob/master/src/main/perf/IndexAndSearchOpenStreetMaps1D.java. Instructions on how to get the data are available at the top of https://github.com/mikemccand/luceneutil/blob/master/src/main/perf/IndexAndSearchOpenStreetMaps.java. It indexes lat,lon data from `geonames` as 1D points. It's not the most realistic benchmark (NYC taxis would be more realistic, but it's also much more complicated to set up) but should give us a good idea of how much better/worse things might get with a dataset that consists mostly of 1D points.

neoremind commented:
I spent some time trying to use the real case benchmark. The speedup of `IndexWriter` is what we expected, faster than main branch, total time elapsed (include adding doc, building index and merging) decreased by about 20%. If we only consider `flush_time`, the speedup is more obvious, time cost drops about 40% - 50%.

1) Run [IndexAndSearchOpenStreetMaps1D.java](https://github.com/neoremind/luceneutil/blob/master/src/main/perf/IndexAndSearchOpenStreetMaps1D.java) against the two branches and take down the [log](https://github.com/neoremind/luceneutil/tree/master/log/OpenStreetMaps).
_note: comment query stage, modify some of the code to adapt to latest Lucene main branch._

main branch:
```
# egrep "flush time|sec to build index" open-street-maps.log
DWPT 0 [2021-04-27T11:33:04.518908Z; main]: flush time 17284.537739 msec
DWPT 0 [2021-04-27T11:33:37.888449Z; main]: flush time 12039.476885 msec
72.49147722 sec to build index
```
PR branch:
```
#egrep "flush time|sec to build index" open-street-maps-optimized.log
DWPT 0 [2021-04-27T11:35:00.619683Z; main]: flush time 9313.007647 msec
DWPT 0 [2021-04-27T11:35:29.575254Z; main]: flush time 8631.820226 msec
59.252797133 sec to build index
```

2) Further more, I come up with an idea to use TPC-H LINEITEM to verify. I have a 10GB TPC-H dataset and develop a new test case to import the first 5 INT fields, which is more typical in real case.

Run [IndexAndSearchTpcHLineItem.java](https://github.com/neoremind/luceneutil/blob/master/src/main/perf/IndexAndSearchTpcHLineItem.java) against the two branches and take down the [log](https://github.com/neoremind/luceneutil/tree/master/log/TPC-H-LINEITEM).

main branch:
```
egrep "flush time|sec to build index" tpch-lineitem.log
DWPT 0 [2021-04-27T11:17:25.329006Z; main]: flush time 13850.23328 msec
DWPT 0 [2021-04-27T11:17:50.289370Z; main]: flush time 12228.723665 msec
DWPT 0 [2021-04-27T11:18:15.546002Z; main]: flush time 12537.085005 msec
DWPT 0 [2021-04-27T11:18:40.140413Z; main]: flush time 11819.225223 msec
DWPT 0 [2021-04-27T11:19:04.850989Z; main]: flush time 12004.380921 msec
DWPT 0 [2021-04-27T11:19:29.435183Z; main]: flush time 11850.273453 msec
DWPT 0 [2021-04-27T11:19:54.016951Z; main]: flush time 11882.316067 msec
DWPT 0 [2021-04-27T11:20:18.932727Z; main]: flush time 12223.151464 msec
DWPT 0 [2021-04-27T11:20:43.522117Z; main]: flush time 11871.276323 msec
DWPT 0 [2021-04-27T11:20:52.060300Z; main]: flush time 3422.434221 msec
271.188917715 sec to build index
```
PR branch:
```
 egrep "flush time|sec to build index" tpch-lineitem-optimized.log
DWPT 0 [2021-04-27T11:24:00.362128Z; main]: flush time 7573.05091 msec
DWPT 0 [2021-04-27T11:24:19.498948Z; main]: flush time 7355.376016 msec
DWPT 0 [2021-04-27T11:24:38.602117Z; main]: flush time 7287.306154 msec
DWPT 0 [2021-04-27T11:24:57.541930Z; main]: flush time 7227.514396 msec
DWPT 0 [2021-04-27T11:25:16.474158Z; main]: flush time 7236.208865 msec
DWPT 0 [2021-04-27T11:25:35.339855Z; main]: flush time 7152.876269 msec
DWPT 0 [2021-04-27T11:25:54.111110Z; main]: flush time 7080.405571 msec
DWPT 0 [2021-04-27T11:26:12.985489Z; main]: flush time 7188.012278 msec
DWPT 0 [2021-04-27T11:26:31.857053Z; main]: flush time 7176.303704 msec
DWPT 0 [2021-04-27T11:26:38.838771Z; main]: flush time 2185.742347 msec
213.175509249 sec to build index
```

For benchmark command, please refer to [my document](https://github.com/neoremind/luceneutil/tree/master/command). 

Test environment:
```
CPU: 
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                32
On-line CPU(s) list:   0-31
Thread(s) per core:    2
Core(s) per socket:    16
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel(R) Xeon(R) Platinum 8163 CPU @ 2.50GHz
Stepping:              4
CPU MHz:               2500.000
BogoMIPS:              5000.00
Hypervisor vendor:     KVM
Virtualization type:   full
L1d cache:             32K
L1i cache:             32K
L2 cache:              1024K
L3 cache:              33792K
NUMA node0 CPU(s):     0-31

Memory: 
$cat /proc/meminfo
MemTotal:       65703704 kB

Disk: SATA 
$fdisk -l | grep Disk
Disk /dev/vdb: 35184.4 GB, 35184372088832 bytes, 68719476736 sectors

OS: 
Linux 4.19.57-15.1.al7.x86_64

JDK:
openjdk version "11.0.11" 2021-04-20 LTS
OpenJDK Runtime Environment 18.9 (build 11.0.11+9-LTS)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.11+9-LTS, mixed mode, sharing)
```

jpountz commented on lucene/core/src/test/org/apache/lucene/util/bkd/TestBKD.java:
```diff
@@ -1698,6 +1698,16 @@ public void getValue(int i, BytesRef packedValue) {
+          @Override
+          public void finalizeAssign(int from, int to) {
+            // do nothing
+          }
```
can you throw an UnexpectedOperationException in both these methods instead, like we do for other methods?

neoremind replied:
comment addressed.

jpountz commented on lucene/core/src/java/org/apache/lucene/util/StableMSBRadixSorter.java:
```diff
@@ -0,0 +1,60 @@
+ *
+ * @lucene.internal
+ */
+public abstract class StableMSBRadixSorter extends MSBRadixSorter {
```
Since the name has "stable", can you override `getFallbackSorter` to return an `InPlaceMergeSorter` so that the sort is guaranteed to be stable in all cases?

neoremind replied:
The `InPlaceMergeSorter` works like a closure, it needs some runtime variables like `reader` and `config` to construct, so it is not able to override `getFallbackSorter`. How about make `getFallbackSorter` throws an UnexpectedOperationException here explicitly, to give the caller a hint to override?

jpountz replied:
I'm confused, wouldn't something like that not work?

```
--- a/lucene/core/src/java/org/apache/lucene/util/StableMSBRadixSorter.java
+++ b/lucene/core/src/java/org/apache/lucene/util/StableMSBRadixSorter.java
@@ -51,4 +51,19 @@ public abstract class StableMSBRadixSorter extends MSBRadixSorter {
     }
     restore(from, to);
   }
+
+  @Override
+  protected Sorter getFallbackSorter(int k) {
+    return new InPlaceMergeSorter() {
+      @Override
+      protected void swap(int i, int j) {
+        StableMSBRadixSorter.this.swap(i, j);
+      }
+
+      @Override
+      protected int compare(int i, int j) {
+        return StableMSBRadixSorter.this.compare(i, j);
+      }
+    };
+  }
 }
```

neoremind replied:
To make it work, there needs a new abstract method in `StableMSBRadixSorter` to compare from the k-th offset in a byte array. Can I create a method like below?
```
protected abstract int compare(int k, int i, int j)
```

jpountz replied:
I don't mind adding one if we give it a default implementation that leverages `byteAt`.

jpountz replied:
Or we could just hardcode comparison using `byteAt` like the fallback IntroSorter does today on MSBRadixSorter.

jpountz reviewed (commented):
Thanks for the additional iterations and benchmarks, this is great. I left some comments but I think we can merge soon.

neoremind committed:
address some comments

neoremind commented on lucene/core/src/java/org/apache/lucene/codecs/MutablePointValues.java:
```diff
@@ -41,4 +41,10 @@ protected MutablePointValues() {}
+  public abstract void assign(int from, int to);
+
+  /** Finalize assign operation, to switch array. */
+  public abstract void finalizeAssign(int from, int to);
```
I have updated the two method names. In terms of performance, `memcpy` should be faster than many `memset` operations, so I thought maybe the current implementation makes sense, what do you think?

jpountz replied:
Sorry I was probably not clear. My point was that the current implementation does this:
 - reorder from the main array into a temporary array
 - copy the content of the temporary array back into the main array

I'm suggesting doing this instead:
 - copy the content of the main array into the temporary array
 - reorder from the temporary array into the main array

I don't think it should make any different in terms of performance, the only benefit I'm seeing is that it would allow us to have a more consistent API with TimSorter. What do you think?

neoremind replied:
I see what you mean, my approach is to reordering then copying, while your approach goes in reverse way. I will try refactor.

jpountz replied:
This sounds good to me, thanks for looking into this!

neoremind commented on lucene/core/src/java/org/apache/lucene/util/StableMSBRadixSorter.java:
```diff
@@ -0,0 +1,60 @@
+ */
+public abstract class StableMSBRadixSorter extends MSBRadixSorter {
+
+  protected boolean useStableSort;
```
removed

neoremind commented on lucene/core/src/java/org/apache/lucene/util/StableMSBRadixSorter.java:
```diff
@@ -0,0 +1,60 @@
+  /** Finalize assign operation, to switch array. */
+  protected void finalizeAssign(int from, int to) {
+    throw new UnsupportedOperationException();
+  }
```
comment addressed.

neoremind commented on lucene/core/src/java/org/apache/lucene/util/bkd/MutablePointsReaderUtils.java:
```diff
@@ -35,63 +37,60 @@
+   * Sort the given {@link MutablePointValues} based on its packed value, note that doc ID is not
+   * taken into sorting algorithm, since if they are already in ascending order, stable sort is able
+   * to maintain the ordering of doc ID.
+   */
```
Sure, I can check here, well, I search the codebase, the only user is `BKDWriter`, the others are some testcases (actually I updated `TestMutablePointsReaderUtils` to ignore doc ID when comparing), could we make it a must for the input to have doc IDs in ascending order?

neoremind replied:
What if the doc IDs came in are not in ascending order? Shall we use MSBRadixSorter with stable reorder to sort? If so, then there needs an if-else clause to provide two different sorters.

neoremind replied:
@jpountz any further suggestions?

jpountz replied:
Oh sorry I had missed your comments. The code path I'm concerned about is this one https://github.com/apache/lucene/blob/d03662c48bfc5bf2be4840a7f743f9cb64b17fee/lucene/core/src/java/org/apache/lucene/index/PointValuesWriter.java#L178-L201 when sortMap is not null, which ends up calling MutablePointsReaderUtil.sort. In practice sortMap is not null when index sorting is configured.

I guess we have 3 options:
 1. Make MutablePointsReaderUtils check whether doc IDs are ordered first, so that it knows whether to use doc IDs when sorting.
 2. Change PointValuesWriter to sort by doc ID before calling MutablePointsReaderUtil.sort when index sorting is enabled.
 3. Change PointValuesWriter to do a second pass over results after calling `MutablePointsReaderUtil.sort` to sort identical values by doc ID when index sorting is enabled.

I suggested (1) initially and your suggestion is (2), which might be a good idea actually if we can reuse the sortMap to reorder the `MutablePointsReader` by doc ID without actually having to sort (ie. O(n) rather than O(n log n)?

jpountz replied:
You should be able to use `sortMap#oldToNew` on doc IDs to know what their index should be if sorted by increasing doc ID (before the wrapping happens, ie. right before [this line](https://github.com/apache/lucene/blob/d03662c48bfc5bf2be4840a7f743f9cb64b17fee/lucene/core/src/java/org/apache/lucene/index/PointValuesWriter.java#L181)).

neoremind committed:
Add fallback sorter in StableMSBRadixSorter, make it work for index sorting.

jpountz committed:
Handle index sorting.

jpountz reviewed (commented):
Sorry for the back and forth about index sorting but I don't think that this approach works as I'm getting some test failures. I made a proposal how to move forward at https://github.com/neoremind/lucene/pull/1. I think we could merge your PR soon with a simple and safe approach like that and see in a follow-up if we can make this even more efficient?

neoremind committed:
Merge pull request #1 from jpountz/lucene9932

Handle index sorting.

neoremind commented on lucene/core/src/java/org/apache/lucene/util/StableMSBRadixSorter.java:
```diff
@@ -0,0 +1,78 @@
+      @Override
+      protected int compare(int i, int j) {
+        for (int o = k; o < maxLength; ++o) {
+          final int b1 = byteAt(i, o);
```
I saw the override method in `MutablePointsReaderUtils` already implemented to be compatible with negative bytes?
```
@Override
      protected int byteAt(int i, int k) {
        if (k < config.packedBytesLength) {
          return Byte.toUnsignedInt(reader.getByteAt(i, k));
        } else {
          final int shift = bitsPerDocId - ((k - config.packedBytesLength + 1) << 3);
          return (reader.getDocID(i) >>> Math.max(0, shift)) & 0xff;
        }
      }
```

jpountz replied:
Thanks, I was confused. :)

neoremind committed:
Update trivial edge case in test case.

neoremind commented on lucene/core/src/test/org/apache/lucene/util/bkd/TestMutablePointsReaderUtils.java:
```diff
@@ -180,15 +189,16 @@ private static BKDConfig createRandomConfig() {
         random().nextBytes(value);
-        points[i] = new Point(value, random().nextInt(maxDoc));
+        points[i] =
+            new Point(value, isDocIdIncremental ? Math.min(i, maxDoc) : random().nextInt(maxDoc));
```
Good catch!

neoremind commented:
Thanks for taking time working on my branch. I merged your change into this PR, the code looks much better 😄.
I was wondering which test case do I neglect besides `TestIndexSorting`? I agree that maybe we could work on improving index sorting scenario later.

jpountz commented:
Agreed, this change helped me notice that we are probably lacking coverage for index sorting. Since this is a pre-existing issue, let's work on it in a separate pull request. In my opinion, the two important bits that are important to test with your PR are `MutablePointsReaderUtils` in the case when doc IDs are incremental, which looks covered already, and `StableMSBRadixSorter`, let's add a new test for this one similar to `TestMSBRadixSorter` and with additional checks regarding stability?

jpountz commented on lucene/core/src/java/org/apache/lucene/util/StableMSBRadixSorter.java:
```diff
@@ -0,0 +1,78 @@
+   */
+  @Override
+  protected void reorder(int from, int to, int[] startOffsets, int[] endOffsets, int k) {
+    int[] assignPos = ArrayUtil.copyOfSubArray(startOffsets, 0, startOffsets.length);
```
Could we reuse this array across calls to `reorder`? We could allocate it up-front since it has a known size of `HISTOGRAM_SIZE`?

jpountz reviewed (commented):
I left some minor comments but in my opinion, it's ready to merge. As final steps, can you:
 - add tests for StableMSBRadixSorter
 - remove the benchmark since we don't check in benchmarks in our source tree
 - add an entry to `CHANGES` under `Lucene 8.9.0`

neoremind committed:
Address comments.
1. Reuse offset array in reorder.
2. Update CHANGES document.
3. Remove benchmark testcase.
4. Add check docIds ordering in `TestMutablePointsReaderUtils.java`.
5. Add new test case for `TestStableMSBRadixSorter.java`.

neoremind commented:
Comments addressed.
1. Reuse offset array with size of `HISTOGRAM_SIZE` in reorder.
2. Update CHANGES document.
3. Remove benchmark test case.
4. Add logic to check doc IDs' ordering in `TestMutablePointsReaderUtils.java`.
5. Add new test case for `StableMSBRadixSorter.java` (by copying `TestMSBRadixSorter` and only change the sorter, is that ok?)

jpountz commented:
@neoremind Did you forget to push your latest commit?

No need to squash yourself, I'll do it upon merge via the GitHub UI. :)

neoremind commented:
@jpountz sorry, forget to push :sweat_smile:

jpountz reviewed (commented):
This looks good to me. Can we better check that the sort is actually stable in the tests? E.g. maybe we could verify that the arrays are not only equal after sorting with Arrays#sort and StableMSBRadixSorter but also that they have the very same instance at every index?

neoremind committed:
Check same instance in test.

neoremind commented:
> This looks good to me. Can we better check that the sort is actually stable in the tests? E.g. maybe we could verify that the arrays are not only equal after sorting with Arrays#sort and StableMSBRadixSorter but also that they have the very same instance at every index?

Add a new assertion `assertSame(points[i].packedValue, reader.points[i].packedValue);`.

jpountz commented:
Can we test it in TestStableMSBRadixSorter too? It would be a pity if bugs in StableMSBRadixSorter were only caught by tests for points.

neoremind committed:
Check same instance in test.

neoremind commented:
Comment addressed.

jpountz reviewed (approved):
Thank you! I'll merge soon.

jpountz committed:
Spotless.

jpountz committed:
Merge branch 'main' into main

neoremind commented:
@jpountz It's great to work with you on this optimization :smile: Thanks for taking so much time to help me.

jpountz commented:
@neoremind I enjoyed it too. Thanks for identifying this opportunity for speedup and going through the many feedback iterations.
